### PR TITLE
Use EF for SQLite interfacing

### PR DESCRIPTION
**Pull request by UselessToucan**
Use EF Core for database interaction.

Depends on [this](https://github.com/ppy/osu-framework/pull/1075) framework PR.

### CODE REVIEW

**Comment by peppy** [comment]
Why all the renames though?

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 1]
```diff
@@ -168,7 +168,7 @@ public BeatmapSetInfo Import(ArchiveReader archiveReader)
 
             // let's only allow one concurrent import at a time for now.
             lock (importLock)
-                connection.RunInTransaction(() => Import(set = importToStorage(archiveReader)));
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapManager.cs]: what happened to transaction logic?

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 2]
```diff
@@ -502,16 +492,12 @@ private BeatmapSetInfo importToStorage(ArchiveReader reader)
         /// <summary>
         /// Returns a list of all usable <see cref="BeatmapSetInfo"/>s.
         /// </summary>
-        /// <param name="populate">Whether returned objects should be pre-populated with all data.</param>
         /// <returns>A list of available <see cref="BeatmapSetInfo"/>.</returns>
-        public List<BeatmapSetInfo> GetAllUsableBeatmapSets(bool populate = true)
+        public List<BeatmapSetInfo> GetAllUsableBeatmapSets()
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapManager.cs]: why removing functionality?

File: osu.Game/Database/DatabaseBackedStore.cs [hunk 3]
```diff
@@ -2,27 +2,23 @@
...
 
 namespace osu.Game.Database
 {
     public abstract class DatabaseBackedStore
     {
         protected readonly Storage Storage;
-        protected readonly SQLiteConnection Connection;
+        protected readonly OsuDbContext Connection;
 
-        protected virtual int StoreVersion => 1;
```
Comment (peppy) [inline osu.Game/Database/DatabaseBackedStore.cs]: ????

File: osu.Game/Database/DbContextBase.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿using Microsoft.EntityFrameworkCore;
+
+namespace osu.Game.Database
+{
+    public abstract class DbContextBase:DbContext
```
Comment (peppy) [inline osu.Game/Database/DbContextBase.cs]: what is this

File: osu.Game/Database/OsuDbContext.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+using osu.Game.IO;
+using osu.Game.Rulesets;
+
+namespace osu.Game.Database
+{
+    public class OsuDbContext : DbContext
+    {
+        private readonly string connectionString;
+
+        public OsuDbContext()
```
Comment (peppy) [inline osu.Game/Database/OsuDbContext.cs]: what is this? why are there no comments?

File: osu.Game/Database/OsuDbContext.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
+        {
+            base.OnConfiguring(optionsBuilder);
+            optionsBuilder.UseSqlite(connectionString);
+        }
+
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            base.OnModelCreating(modelBuilder);
+            modelBuilder.Entity<BeatmapInfo>().HasIndex(b => b.MD5Hash);
```
Comment (peppy) [inline osu.Game/Database/OsuDbContext.cs]: no

File: osu.Game/OsuGameBase.cs [hunk 7]
```diff
@@ -82,35 +83,41 @@ public OsuGameBase()
...
             dependencies = new DependencyContainer(base.CreateLocalDependencies(parent));
 
-        private SQLiteConnection createConnection()
+        private OsuDbContext createDbContext()
         {
-            var conn = Host.Storage.GetDatabase(@"client");
-            conn.BusyTimeout = new TimeSpan(TimeSpan.TicksPerSecond * 10);
-            return conn;
+            var connectionString = Host.Storage.GetDatabaseConnectionString(@"client");
+            var context = new OsuDbContext(connectionString);
```
Comment (peppy) [inline osu.Game/OsuGameBase.cs]: this is all stupid. why is this not at the framework level? why is sqlite implemented in osu.game? why is there even anything left in the framework? it's pointless at this point because all the implementation (except for the connection string?) is here.

and it's in OsuGameBase

no please god no

File: osu.Game/Beatmaps/BeatmapManager.cs (hunk 1 above)
Comment (UselessToucan) [reply osu.Game/Beatmaps/BeatmapManager.cs]: AFAIK SaveChanges method uses transactions under the hood. Also, it is possible to use transactions manually using DbContext.Database.BeginTransaction() method.
https://docs.microsoft.com/en-us/ef/core/saving/transactions

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapManager.cs]: Sure, but from what I can see here you removed the transaction and didn’t leave a todo or anything.

File: osu.Game/Beatmaps/BeatmapManager.cs (hunk 2 above)
Comment (UselessToucan) [reply osu.Game/Beatmaps/BeatmapManager.cs]: Eager loading is implemented in Query methods(https://github.com/UselessToucan/osu/blob/EfCore/osu.Game/Beatmaps/BeatmapStore.cs#L138).
If this is critical for you, I will try to extract it to Populate method.

File: osu.Game/Database/DatabaseBackedStore.cs (hunk 3 above)
Comment (UselessToucan) [reply osu.Game/Database/DatabaseBackedStore.cs]: As far as I understand, this property was used only for DB migrations. In EF we can migrate DB without it. I can just call `dbContext.Database.Migrate();` and pending migrations will be applied(if there are any).

File: osu.Game/Database/OsuDbContext.cs (hunk 5 above)
Comment (UselessToucan) [reply osu.Game/Database/OsuDbContext.cs]: My bad. Will add comments this evening when I get back home.

File: osu.Game/OsuGameBase.cs (hunk 7 above)
Comment (UselessToucan) [reply osu.Game/OsuGameBase.cs]: I can't create OsuDbContext at the framework level because it must contain DbSets(which are game specific).

Comment (peppy) [reply osu.Game/OsuGameBase.cs]: if t can’t be abstracted to the framework then it shouldn’t exist in the framework, but i’m pretty sure it can be.

**Comment by peppy** [comment]
Please begin by undoing the naming changes and keeping this PR to the minimal required implementation.

**Comment by UselessToucan** [comment]
I reviewed naming changes and noticed that most of them are Id properties.
It is [EF Convention](https://docs.microsoft.com/en-us/ef/core/modeling/relationships#conventions) for primary/foreign keys.

**Comment by smoogipoo** [comment]
I would rather us temporarily not abiding to conventions if it means not going through a 2000 line change PR.

**Comment by peppy** [comment]
Also what are the .Designer files generated from? We'll probably not want to do that, and generate by hand to fix bad formatting and allow people who aren't using (whatever IDE created them) to still work on the project easily.

**Comment by UselessToucan** [comment]
> Also what are the .Designer files generated from?
20171009205048_Init.cs, 20171009205048_Init.Designer.cs, OsuDbContextModelSnapshot.cs were generated by "Add-Migration <Name>" call. These files are required for creating(or updating) database on the client machine.

File: osu.Game/Beatmaps/BeatmapSetOnlineInfo.cs [hunk 8]
```diff
@@ -12,11 +15,19 @@ namespace osu.Game.Beatmaps
     /// </summary>
     public class BeatmapSetOnlineInfo
     {
+        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
+        public int Id { get; set; }
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapSetOnlineInfo.cs]: why is this added (and why is this class modified)? currently this stuff isn't stored in the database

Comment (UselessToucan) [reply osu.Game/Beatmaps/BeatmapSetOnlineInfo.cs]: Yes, there is actually no need to modify this class. It seems that I overdid it.

File: osu.Game/osu.Game.csproj [hunk 9]
```diff
@@ -116,27 +160,52 @@
...
+      <HintPath>..\packages\System.Runtime.CompilerServices.Unsafe.4.4.0\lib\netstandard2.0\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
+    </Reference>
+    <Reference Include="System.ValueTuple">
+      <HintPath>..\packages\System.ValueTuple.4.4.0\lib\net461\System.ValueTuple.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="SQLiteNetExtensions, Version=1.0.0.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>$(SolutionDir)\packages\SQLiteNetExtensions.1.3.0\lib\portable-net45+netcore45+wpa81+wp8+MonoAndroid1+MonoTouch1\SQLiteNetExtensions.dll</HintPath>
+    <Reference Include="System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51">
+      <HintPath>../packages/System.ValueTuple.4.4.0/lib/net461/System.ValueTuple.dll</HintPath>
```
Comment (peppy) [inline osu.Game/osu.Game.csproj]: / \

File: osu.Game/Beatmaps/BeatmapStore.cs [hunk 10]
```diff
@@ -125,46 +127,20 @@ public bool Restore(BeatmapInfo beatmap)
...
-        public List<BeatmapInfo> QueryBeatmaps(Expression<Func<BeatmapInfo, bool>> query)
-        {
-            return Connection.BeatmapInfo
-                .Include(b => b.BeatmapSet)
-                .Include(b => b.Metadata)
-                .Include(b => b.Ruleset)
-                .Include(b => b.Difficulty)
-                .Where(query).ToList();
-        }
+        public IEnumerable<BeatmapSetInfo> BeatmapSets => Connection.BeatmapSetInfo
```
Comment (UselessToucan) [inline osu.Game/Beatmaps/BeatmapStore.cs]: I like these two properties.
Shame on me because I did not come to this implementation from the very beginning.

**Comment by peppy** [comment]
remaining tasks are

- go through the actual EF database structure, including fixing the cascade deletion to FileInfo and BeatmapDifficulty which may or may not be broken
- reinstate transactions (i've already read into it and it doesn't look to be an issue)
- comprehensive testing of multi-connection interaction (since the new code is switching to provide a db context per store)
- fixing of the fucked up ValueTuple crap which is causing CI tests to fail (this is a really shocking situation, there's hundreds of people having the same issue due to weird versioning across different .net platforms)
- adding more tests for each store type
- considering migration from old db format

**Comment by UselessToucan** [comment]
> fixing the cascade deletion to FileInfo and BeatmapDifficulty which may or may not be broken

@peppy [You reverted this one yourself](https://github.com/ppy/osu/pull/1350/commits/04e5f764a34e807db1a25de355bd79709f3aac74) without explaining what is wrong.

**Comment by peppy** [comment]
i already said i reverted *all your changes*. also that one was *explicitly incorrect* because the foreign key was in the wrong table.

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 11]
```diff
@@ -156,21 +171,39 @@ public void Import(params string[] paths)
...
-            BeatmapSetInfo set = null;
-
             // let's only allow one concurrent import at a time for now.
-            lock (importLock)
-                connection.RunInTransaction(() => Import(set = importToStorage(archiveReader)));
+            lock (importContext)
+            {
+                var context = importContext.Value;
+
+                context.Database.AutoTransactionsEnabled = false;
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/BeatmapManager.cs]: Is this needed? Why is this not needed inside Delete()?

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapManager.cs]: to be honest i'm not sure this has much of a benefit; requires further testing. basically it runs `DetectChanges` less than otherwise would be. When I get to addressing transactions via a better pattern it will likely be automatically toggled if there are benefits.

If you'd prefer we can remove it for now. Or add it to deletion.

Comment (smoogipoo) [reply osu.Game/Beatmaps/BeatmapManager.cs]: Also why is this not being reset back to true?

Comment (peppy) [reply osu.Game/Beatmaps/BeatmapManager.cs]: oh, ignore previous comment, i misread which bool this was. it needs to be set for deletions too.

**Review by smoogipoo — DISMISSED** [review]
This looks pretty good. I've got not major issues with it, beyond that it needs to be abstracted super hard, and needs to be moved to the framework.

File: osu.Game/Database/DatabaseContextFactory.cs [hunk 12]
```diff
@@ -14,6 +14,6 @@ public DatabaseContextFactory(GameHost host)
             this.host = host;
         }
 
-        public OsuDbContext GetContext() => new OsuDbContext(host.Storage.GetDatabaseConnectionString(@"client-ef"));
+        public OsuDbContext GetContext() => new OsuDbContext(host.Storage.GetDatabaseConnectionString(@"client"));
```
Comment (smoogipoo) [inline osu.Game/Database/DatabaseContextFactory.cs]: So what happens when I run TestCasePlaySongSelect?

File: osu.Game/Database/OsuDbContext.cs [hunk 13]
```diff
@@ -156,5 +156,74 @@ public bool IsEnabled(LogLevel logLevel)
...
+
+                    Database.ExecuteSqlCommand("INSERT INTO RulesetInfo SELECT ID, Available, InstantiationInfo, Name FROM RulesetInfo_Old");
+                    Database.ExecuteSqlCommand("DROP TABLE RulesetInfo_Old");
+
+                    Database.ExecuteSqlCommand(
+                        "INSERT INTO BeatmapInfo SELECT ID, AudioLeadIn, BaseDifficultyID, BeatDivisor, BeatmapSetInfoID, Countdown, DistanceSpacing, GridSize, Hash, Hidden, LetterboxInBreaks, MD5Hash, NULLIF(BeatmapMetadataID, 0), OnlineBeatmapID, Path, RulesetID, SpecialStyle, StackLeniency, StarDifficulty, StoredBookmarks, TimelineZoom, Version, WidescreenStoryboard FROM BeatmapInfo_Old");
+                    Database.ExecuteSqlCommand("DROP TABLE BeatmapInfo_Old");
+                }
+                catch
+                {
```
Comment (smoogipoo) [inline osu.Game/Database/OsuDbContext.cs]: Is this going to be a problem if this fails on an older DB?

File: osu.Game/Database/DatabaseContextFactory.cs (hunk 12 above)
Comment (peppy) [reply osu.Game/Database/DatabaseContextFactory.cs]: Nothing has changed there; it uses a facade storage which resolves to an in-memory SQLite instance

**Review by smoogipoo — APPROVED** [review]
